### PR TITLE
Resolving errors with empty data frames

### LINKED ISSUE #402: Dnatracing division by zero
**Describe the bug**
When processing the following file, TopoStats throws a division by zero error during `dnatracing`. 

**To Reproduce**

Run topostats on the following file and with the following config:

File: https://drive.google.com/drive/folders/1mFtdYHl6g0kxxBpJcpY9a65t1TP-kqyU

Config:

```
base_dir: ./raphael/20220819/C3
output_dir: ./output/20220819/C3
warnings: ignore
cores: 4
quiet: false
file_ext: .spm
loading:
  channel: Height
filter:
  run: true # Options : true, false
  threshold_method: std_dev # Options : otsu, std_dev, absolute
  otsu_threshold_multiplier: 1.0
  threshold_std_dev:
    lower: 10.0
    upper: 1.0
  threshold_absolute:
    lower: -1.0
    upper: 1.0
  gaussian_size: 1.0121397464510862 # in px
  gaussian_mode: nearest
grains:
  run: true # Options : true, false
  # Thresholding by height
  threshold_method: std_dev # Options : std_dev, otsu, absolute
  otsu_threshold_multiplier: 1.0
  threshold_std_dev:
    lower: 10.0
    upper: 1.0
  threshold_absolute:
    lower: -1.0
    upper: 1.0
  direction: upper # Options: upper, lower, both (defines whether to look for grains above or below thresholds or both)
  # Thresholding by area
  absolute_smallest_grain_size: 100
  absolute_area_threshold:
    upper: [ 300, 100000 ] # above surface [Low, High] in nm^2 (also takes null)
    lower: [ null, null ] # below surface [Low, High] in nm^2 (also takes null)
grainstats:
  run: true # Options : true, false
  cropped_size: 40.0 # Length (in nm) of square cropped images (can take -1 for grain-sized box)
  save_cropped_grains: true # Options : true, false
dnatracing:
  run: true # Options : true, false
plotting:
  run: true # Options : true, false
  save_format: png # Options : see https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.savefig.html
  image_set: all  # Options : all, core
  zrange: [-4, 4]  # low and high height range for core images (can take [null, null])
  colorbar: true  # Options : true, false
  axes: true # Options : true, false (due to off being a bool when parsed)
  cmap: nanoscope # Options : nanoscope, afmhot
  histogram_log_axis: false # Options : true, false
```

**Output**

```
❯ run_topostats --config ./my_config.yaml
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Configuration is valid.
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Plotting configuration is valid.
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Configuration file loaded from      : ./my_config.yaml
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Scanning for images in              : raphael/20220819/C3
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Output directory                    : output/20220819/C3
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Looking for images with extension   : .spm
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Images with extension .spm in raphael/20220819/C3 : 1
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Thresholding method (Filtering)     : std_dev
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Thresholding method (Grains)        : std_dev
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Extracting image from raphael/20220819/C3/C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes.0_00013.spm
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] Loading image from : raphael/20220819/C3/C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes.0_00013.spm
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes.0_00013] : Loaded image from : raphael/20220819/C3/C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes.0_00013.spm
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes.0_00013] : Extracted channel Height
[Tue, 13 Dec 2022 18:50:16] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes.0_00013] : Pixel to nm scaling : 7.8125
Processing images from raphael/20220819/C3, results are under output/20220819/C3:   0%|               | 0/1 [00:00<?, ?it/s][Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] image_path: raphael/20220819/C3/C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes.0_00013
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] image_path: raphael/20220819/C3/C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] Processing : C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] NO SUFFIX : 
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] output_dir: output/20220819/C3
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] _output_dir: output/20220819/C3/Processed
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] processed path: output/20220819/C3/Processed
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] folder path: raphael/20220819/C3
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats]  filter out path: output/20220819/C3/Processed/C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013/filters
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats]  grains out path: output/20220819/C3/Processed/C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013/grains
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Median flattening without mask
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Plane tilt removal without mask
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] x-polyfit 1st order: [  0.27207554 -17.41082201]
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] y-polyfit 1st order: [-1.43326723e-18  5.07346103e-16]
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] removing x plane tilt
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] removing y plane tilt
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Remove quadratic bow without mask
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] x polyfit 2nd order: [-8.73271071e-05  1.10905426e-02 -1.76437234e+01]
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Masking (upper) Threshold: -16.75327140083654
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Masking (lower) Threshold: -21.95389593854906
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Median flattening with mask
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Plane tilt removal with mask
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] x-polyfit 1st order: [ 4.20504532e-06 -1.08544093e-02]
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] y-polyfit 1st order: [-1.60607249e-18  1.29741179e-16]
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] removing x plane tilt
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] removing y plane tilt
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Remove quadratic bow with mask
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] x polyfit 2nd order: [-8.14653855e-05  1.03461040e-02 -2.28122592e-01]
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Zero averaging background : 0.1146772811694596 nm
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Applying Gaussian filter (mode : nearest; Gaussian blur (px) : 1.0121397464510862).
[Tue, 13 Dec 2022 18:50:17] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Plotting Filtering Images
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Image saved to : output/20220819/C3/Processed/C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013_processed
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : *** Grain Finding ***
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Thresholding method (grains) : std_dev
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Processing upper threshold (0.34804475650010547)
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Masking (upper) Threshold: 0.34804475650010547
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Labelling Regions
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Tidying borders
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Labelling Regions
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Region properties calculated (upper)
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Coloured regions
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Extracted bounding boxes (upper)
[Tue, 13 Dec 2022 18:50:20] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Plotting Grain Finding Images
[Tue, 13 Dec 2022 18:50:21] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013_upper] : Image saved to : output/20220819/C3/Processed/C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013_upper_processed_masked
[Tue, 13 Dec 2022 18:50:21] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : *** Grain Statistics ***
[Tue, 13 Dec 2022 18:50:21] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Processing grain: 0
[Tue, 13 Dec 2022 18:50:21] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Processing grain: 1
[Tue, 13 Dec 2022 18:50:21] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Processing grain: 2
[Tue, 13 Dec 2022 18:50:22] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Processing grain: 3
[Tue, 13 Dec 2022 18:50:22] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Processing grain: 4
[Tue, 13 Dec 2022 18:50:22] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : *** DNA Tracing ***
[Tue, 13 Dec 2022 18:50:22] [INFO    ] [topostats] [C3_1ng_Imob_in_MgCl2_3x_NiCl2_washes_0_00013] : Gaussian filter applied.
Processing images from raphael/20220819/C3, results are under output/20220819/C3:   0%|               | 0/1 [00:06<?, ?it/s]
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/Users/sylvi/opt/anaconda3/envs/scars/lib/python3.10/multiprocessing/pool.py", line 125, in worker
    result = (True, func(*args, **kwds))
  File "/Users/sylvi/Documents/GitKraken/TopoStats/topostats/run_topostats.py", line 352, in process_scan
    dna_traces[direction].trace_dna()
  File "/Users/sylvi/Documents/GitKraken/TopoStats/topostats/tracing/dnatracing.py", line 91, in trace_dna
    self.get_splined_traces()
  File "/Users/sylvi/Documents/GitKraken/TopoStats/topostats/tracing/dnatracing.py", line 426, in get_splined_traces
    while nbr / step_size < 4:
ZeroDivisionError: division by zero
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/sylvi/opt/anaconda3/envs/scars/bin/run_topostats", line 8, in <module>
    sys.exit(main())
  File "/Users/sylvi/Documents/GitKraken/TopoStats/topostats/run_topostats.py", line 459, in main
    for img, result in pool.imap_unordered(
  File "/Users/sylvi/opt/anaconda3/envs/scars/lib/python3.10/multiprocessing/pool.py", line 873, in next
    raise value
ZeroDivisionError: division by zero
```

**(Not required if using Docker) Your computer configuration (please complete the following information):**
 - OS: MacOS
 - topostats version: 2.0.0
 - Python version: 3.10

### PR DESCRIPTION
**Pull request by ns-rse**
Closes #402

Still get errors with DNA tracing of this image but it runs to completion.

Changes include...

* Changes to where indexes are set and ensuring `image` is part of pd.DataFrame index.
* Return `grainstats_df` if `tracingdna` fails or `dnatracing_config["run"] == False`.
* Improves flexibility of `create_empty_dataframe()` allowing index to be set explicitly.
* Only attempt to plot summary distribution if statistic has been calculated (required because configs may request plotting of statistics from DNA tracing which is more likely to fail, yet we now return grainstats even if DNA tracing does fail and some of those may be plotted).
* Introduces logic such that stages subsequent to `Filters()` can be disabled without causing errors (although I'm not sure this is something which belongs in TopoStats, it should try and process an image as far as possible then fail, and move onto the next image). EDIT : See next comment in this PR too.
* Move completion message to processing module (includes reporting version).
* Removes duplicated file-extension from logging info when saving distribution plots.
* Updates regression test files with changed columns and grainstats with index name (also log of successful run to reflect footnote introduced in completion message).

### CODE REVIEW

**Comment by ns-rse** [comment]
In light of the desire to toggle certain sections of analyses on/off we need to check that configuration options are logically consistent (i.e. if a later stage in processing is enabled then the required earlier steps are also enabled).

Kind of feels a bit excessive and would be interested to know people's thoughts on having all steps run automatically and directing people to using Jupyter Notebooks (see #242 for more information and branch `ns-rse/242a-notebooks`) if they wish to run only some steps of the processing.

**Comment by SylviaWhittle** [comment]
> In light of the desire to toggle certain sections of analyses on/off we need to check that configuration options are logically consistent (i.e. if a later stage in processing is enabled then the required earlier steps are also enabled).
> 
> Kind of feels a bit excessive and would be interested to know people's thoughts on having all steps run automatically and directing people to using Jupyter Notebooks (see #242 for more information and branch `ns-rse/242a-notebooks`) if they wish to run only some steps of the processing.

This is an interesting problem. I personally like to be able to turn sections off in the interests of speed, as I know a couple experimentalists like to do also. Assuming we speed things up enough when we eventually get around to performance improvements, this is not a use case we should necessarily spend time on.

Are there any workflows that require turning off parts? I do not know. I shall ask tomorrow at the group meeting. I do know that with some silicon wafer `.spm` files, we only want to be applying just the `plane_tilt_removal` within `filters.py`, however this is extremely granular control that I'm not sure is something that we want to burden ourselves with supporting, and hence perhaps that use case should be covered with a notebook rather than the main software?

File: topostats/processing.py [hunk 1]
```diff
@@ -247,16 +255,120 @@ def process_scan(
...
-                    # NB - Merge on molecule and threshold because we may have upper and lower molecueles which gives
-                    #      duplicate molecule numbers as they are processed separately
-                    results = grainstats_df.merge(tracing_stats_df, on=["molecule_number", "threshold"])
+                    # NB - Merge on image, molecule and threshold because we may have upper and lower molecueles which
+                    #      gives duplicate molecule numbers as they are processed separately
+                    results = grainstats_df.merge(tracing_stats_df, on=["image", "threshold", "molecule_number"])
                 else:
+                    LOGGER.info(f"[{filename}] Calculation of DNA Tracing disabled, returning grainstats data frame.")
                     results = grainstats_df
                     results["image"] = filename
```
Comment (SylviaWhittle) [inline topostats/processing.py:264]: Might want to remove this as we found out it is redundant earlier? I think you may have already done this locally.

File: topostats/processing.py [hunk 2]
```diff
@@ -247,16 +255,120 @@ def process_scan(
...
             except Exception:
                 # If no results we need a dummy dataframe to return.
-                LOGGER.info(f"[{filename}] : Errors occurred whilst calculating DNA tracing statistics.")
-                results = create_empty_dataframe()
+                LOGGER.info(
+                    f"[{filename}] : Errors occurred whilst calculating DNA tracing statistics, "
+                    "returning grain statistics"
+                )
+                results = grainstats_df
+                results["image"] = filename
```
Comment (SylviaWhittle) [inline topostats/processing.py:273]: Likewise here, a possibly redundant setting of the `"image"` parameter

File: topostats/processing.py [hunk 3]
```diff
@@ -247,16 +255,120 @@ def process_scan(
...
+            LOGGER.info(f"[{filename}] Calculation of grainstats disabled, returning empty data frame.")
+            results = create_empty_dataframe()
+    else:
+        LOGGER.info(f"[{filename}] Detection of grains disabled, returning empty data frame.")
+        results = create_empty_dataframe()
 
     return image_path, results
+
+
+def check_run_steps(filter_run: bool, grains_run: bool, grainstats_run: bool, dnatracing_run: bool) -> None:
```
Comment (SylviaWhittle) [inline topostats/processing.py:285]: Excellent idea to add this!

**Review by SylviaWhittle — CHANGES_REQUESTED** [review]
Really useful changes, just a couple things we already discussed.

Checked that combinations of disabling / enabling worked as intended / expected 👍

File: topostats/processing.py (hunk 1 above)
Comment (ns-rse) [reply topostats/processing.py:264]: D'oh, I forgot to push the removals last night, just pushed them.

**Comment by ns-rse** [comment]
> Checked that combinations of disabling / enabling worked as intended / expected

Excellent thanks for checking that. I wasn't sure if the tests correctly covered things so its good to have a third party check.

**Review by SylviaWhittle — APPROVED** [review]
Loving how as the code is becoming more modular, making structural changes is easier and makes more sense.

The tests you have added are very elegant.

**Pull request merged**